src/helics/core/mpi/MpiComms.cpp

```diff
@@ -9,38 +9,319 @@ Lawrence Livermore National Laboratory, operated by Lawrence Livermore National
 #include <memory>
 
+#include <mpi.h>
+
 namespace helics
 {
-MpiComms::MpiComms (const std::string &brokerTarget, const std::string &localTarget)
-    : CommsInterface (brokerTarget, localTarget)
+
+bool MpiComms::mpiCommsExists = false;
+std::mutex MpiComms::mpiSerialMutex;
+int MpiComms::commRank = -1;
+
+MpiComms::MpiComms ()
 {
+    if (mpiCommsExists)
+    {
+        std::cout << "WARNING: MPIComms object already created, unexpected results may occur" << std::endl;
+    }
+    std::cout << "MpiComms()" << std::endl;
+    initMPI();
+    mpiCommsExists = true;
+}
+
+MpiComms::MpiComms(const int &brokerRank)
+    : brokerRank(brokerRank)
+{
+    if (mpiCommsExists)
+    {
+        std::cout << "WARNING: MPIComms object already created, unexpected results may occur" << std::endl;
+    }
+    std::cout << "MpiComms(" << std::to_string(brokerRank) << ")" << std::endl;
+    initMPI();
+    if (brokerRank == -1)
+    {
+        setBrokerRank (commRank);
+    }
+    mpiCommsExists = true;
 }
+
 /** destructor*/
-MpiComms::~MpiComms () = default;
+MpiComms::~MpiComms ()
+{
+    std::cout << "Disconnecting MPIComms" << std::endl;
+    disconnect ();
+    std::cout << "Finished disconnect" << std::endl;
 
-void MpiComms::queue_rx_function () {}
+    int mpi_initialized;
+    MPI_Initialized(&mpi_initialized);
+    if (mpi_initialized)
+    {
+        std::cout << "Finalizing MPI" << std::endl;
+        std::lock_guard<std::mutex> lock(mpiSerialMutex);
+        MPI_Finalize();
+        shutdown = true;
+        std::cout << "MPI Finalized" << std::endl;
+    }
+}
 
-void MpiComms::closeTransmitter ()
+bool MpiComms::initMPI()
 {
-    ActionMessage rt (CMD_PROTOCOL);
-    rt.index = DISCONNECT;
-    transmit (-1, rt);
+    std::lock_guard<std::mutex> lock(mpiSerialMutex);
+
+    // Initialize MPI with MPI_THREAD_SERIALIZED
+    int mpi_initialized;
+    int mpi_thread_level;
+
+    MPI_Initialized(&mpi_initialized);
+
+    if (!mpi_initialized)
+    {
+        MPI_Init_thread(nullptr, nullptr, MPI_THREAD_SERIALIZED, &mpi_thread_level);
+
+        MPI_Initialized(&mpi_initialized);
+        if (!mpi_initialized)
+        {
+            std::cerr << "MPI initialization failed" << std::endl;
+            return false;
+        }
+    }
+
+    MPI_Query_thread(&mpi_thread_level);
+
+    if (mpi_thread_level < MPI_THREAD_SERIALIZED)
+    {
+        std::cerr << "MPI_THREAD_SERIALIZED support required" << std::endl;
+        return false;
+    }
+
+    return true;
 }
 
-void MpiComms::closeReceiver ()
+void MpiComms::serializeSendMPI(std::vector<char> message, int dest, int tag, MPI_Comm comm) {
+    std::cout << "SendMPI from " << getAddress() << " to " << dest << std::endl;
+    std::lock_guard<std::mutex> lock(mpiSerialMutex);
+    MPI_Request req;
+    MPI_Isend(message.data(), message.size(), MPI_CHAR, dest, tag, comm, &req);
+
+    int message_sent = false;
+    while (!message_sent)
+    {
+        MPI_Test(&req, &message_sent, MPI_STATUS_IGNORE);
```

Comment:
Is there any chance this gets into an infinite loop?

Reply:
There is, if the send request never completes.